packages/menu/src/Menu.tsx

```diff
@@ -35,87 +52,114 @@ interface MenuProps extends Omit<PopoverProps, 'spacing'> {
  */
+
+const escapeKey = 27;
+
 function Menu({
   align = Align.Bottom,
   justify = Justify.End,
   usePortal = true,
   adjustOnMutation = false,
-  active,
+  open = false,
+  shouldClose = () => true,
+  setOpen,
   children,
   className,
   refEl,
   trigger,
   ...rest
 }: MenuProps) {
-  const [isActive, setActiveState] = useState(false);
+  const [uncontrolledOpen, setUncontrolledOpen] = useState(false);
+
+  const popoverRef: React.RefObject<HTMLDivElement> = useRef(null);
+
+  const handleClose = () => {
+    if (setOpen && shouldClose()) {
+      setOpen(false);
+    }
+
+    if (!setOpen && shouldClose()) {
+      setUncontrolledOpen(false);
+    }
+  };
+
+  const handleBackdropClick = (e: MouseEvent) => {
+    const popoverReference = popoverRef && popoverRef.current;
+
+    if (
+      popoverReference &&
+      !popoverReference.contains(e.target as HTMLElement)
+    ) {
+      handleClose();
+    }
+  };
+
+  const enabled = open || uncontrolledOpen;
+
+  useEventListener('click', handleBackdropClick, {
+    enabled,
+  });
+
+  const handleEscape = (e: KeyboardEvent) => {
+    e.stopImmediatePropagation();
+
+    if (e.keyCode === escapeKey) {
+      handleClose();
+    }
+  };
+
+  useEventListener('keydown', handleEscape, {
+    enabled,
+  });
 
   const popoverContent = (
     <Popover
       key="popover"
-      active={trigger ? isActive : active}
+      active={setOpen ? open : uncontrolledOpen}
       align={align}
       justify={justify}
       refEl={refEl}
       usePortal={usePortal}
       spacing={15}
       adjustOnMutation={adjustOnMutation}
     >
-      <div {...rest} className={cx(rootMenuStyle, className)} role="menu">
+      <div
+        {...rest}
+        className={cx(rootMenuStyle, className)}
+        role="menu"
+        ref={popoverRef}
+      >
         {children}
       </div>
     </Popover>
   );
 
-  const syntheticToggleEventHandler: EventHandler<
-    React.SyntheticEvent
-  > = useCallback(e => {
-    e.nativeEvent.stopImmediatePropagation();
-    setActiveState(current => !current);
-  }, []);
-
-  const closeMenuNativeHandler: EventListener = useCallback((e: Event) => {
-    e.stopImmediatePropagation();
-    setActiveState(false);
-  }, []);
-
-  const handleEscape = useCallback((e: KeyboardEvent) => {
-    if (e.keyCode === EscapeKey) {
-      closeMenuNativeHandler(e);
-    }
-  }, []);
-
-  const enabled = trigger && isActive;
-
-  useEventListener('click', closeMenuNativeHandler, {
-    options: { once: true },
-    dependencies: [isActive, trigger],
-    enabled,
-  });
-
-  useEventListener('keydown', handleEscape, {
-    options: { once: true },
-    dependencies: [isActive, trigger],
-    enabled,
-  });
-
   if (trigger) {
     if (typeof trigger === 'function') {
       return trigger({
-        onClick: syntheticToggleEventHandler,
+        onClick: () => setUncontrolledOpen(open => !open),
```

Comment:
Hmm... shouldn't this use `uncontrolledOpen` instead of `open`?

Reply:
Yep.... great catch